### PR TITLE
gitserver: Sync soft deleted repos clone status

### PR DESCRIPTION
**Pull request by ryanslade**
We now sync the clone status of repos even after they've been soft deleted in our database. 

This ensures that we accurately list whether or not a deleted repo is still cloned on disk.

## Test plan

Added unit tests. Will also monitor [this metric](https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Prometheus%22,%7B%22exemplar%22:true,%22expr%22:%22max(src_repoupdater_purgeable_repos)%22,%22requestId%22:%22Q-e397beff-8412-498c-a3ef-d9b56a13a5af-0A%22%7D%5D) when we roll out, we expect to see it drop.

### CODE REVIEW

File: cmd/gitserver/server/server.go [hunk 1]
```diff
@@ -427,8 +427,10 @@ func (s *Server) SyncRepoState(interval time.Duration, batchSize, perSecond int)
 		previousAddrs = currentAddrs
 
 		gitServerAddrs := gitserver.GitServerAddresses{
-			Addresses:     addrs,
-			PinnedServers: cfg.ExperimentalFeatures.GitServerPinnedRepos,
+			Addresses: addrs,
+		}
+		if cfg.ExperimentalFeatures != nil {
```
Comment (ryanslade) [inline cmd/gitserver/server/server.go:432]: ExperimentalFeatures could be nil

**Review by mollylogue — APPROVED** [review]
👍

File: cmd/gitserver/server/server.go [hunk 2]
```diff
@@ -595,12 +597,21 @@ func (s *Server) syncRepoState(gitServerAddrs gitserver.GitServerAddresses, batc
 		repoStateUpsertCounter.WithLabelValues("true").Add(float64(len(batch)))
 	}
 
-	options := database.IterateRepoGitserverStatusOptions{}
+	options := database.IterateRepoGitserverStatusOptions{
+		// We also want to include deleted repos as they ay still be cloned on disk
```
Comment (indradhanush) [inline cmd/gitserver/server/server.go:601]: Typo + nit. :grin: 

```suggestion
		// We also want to include deleted repos as they may still be cloned on disk.
```

File: internal/api/api.go [hunk 3]
```diff
@@ -26,6 +28,14 @@ func (r RepoName) Equal(o RepoName) bool {
 	return strings.EqualFold(string(r), string(o))
 }
 
+var deletedRegex = lazyregexp.New("DELETED-[0-9]+\\.[0-9]+-")
+
+// UndeleteRepoName will "undeleted" a repo name. When we soft-delete a repo we
```
Comment (indradhanush) [inline internal/api/api.go:33]:
```suggestion
// UndeleteRepoName will "undelete" a repo name. When we soft-delete a repo we
```
```suggestion
// UndeleteRepoName will "undeleted" a repo name. When we soft-delete a repo we
```

File: internal/api/api.go [hunk 4]
```diff
@@ -26,6 +28,14 @@ func (r RepoName) Equal(o RepoName) bool {
...
+// change its name in the database, this function extract the original repo name.
+func UndeleteRepoName(name RepoName) RepoName {
```
Comment (indradhanush) [inline internal/api/api.go:35]: Either the above, or maybe you wanted to go for `UndeletedRepoName`? 

`UndeleteRepoName` sounds like a more persistent state change vs `UndeletedRepoName`.  Maybe it's just me overthinking. :thinking: 

Feel free to ignore.

```suggestion
func UndeletedRepoName(name RepoName) RepoName {
```

**Review by indradhanush — APPROVED** [review]

File: internal/api/api.go (hunk 4 above)
Comment (ryanslade) [reply internal/api/api.go:36]: Agree that 'UndeletedRepoName' is better, I'll rename

**Pull request merged**